diff --git a/include/tudat/astro/ground_stations/groundStation.h b/include/tudat/astro/ground_stations/groundStation.h
index f462d8781..fe63646f8 100644
--- a/include/tudat/astro/ground_stations/groundStation.h
+++ b/include/tudat/astro/ground_stations/groundStation.h
@@ -116,6 +116,11 @@ public:
         return transmittingFrequencyCalculator_;
     }
 
+    bool hasFrequencyCalculator( )
+    {
+        return ( !( transmittingFrequencyCalculator_ == nullptr  ) ) ;
+    }
+
     //! Function to set the object used to compute the ground station's transmitting frequency at a given time
     void setTransmittingFrequencyCalculator( std::shared_ptr< StationFrequencyInterpolator > transmittingFrequencyCalculator )
     {
diff --git a/include/tudat/astro/ground_stations/transmittingFrequencies.h b/include/tudat/astro/ground_stations/transmittingFrequencies.h
index 5aaad7dff..8be30da25 100644
--- a/include/tudat/astro/ground_stations/transmittingFrequencies.h
+++ b/include/tudat/astro/ground_stations/transmittingFrequencies.h
@@ -175,52 +175,11 @@ public:
         StationFrequencyInterpolator( ), startTimes_( startTimes ), endTimes_( endTimes ), rampRates_( rampRates ),
         startFrequencies_( startFrequency )
     {
-        // Check if dimensions of all vectors are consistent
-        if( startTimes_.size( ) != endTimes_.size( ) || startTimes_.size( ) != rampRates_.size( ) ||
-            startTimes_.size( ) != startFrequencies_.size( ) )
-        {
-            throw std::runtime_error(
-                    "Error when creating piecewise linear frequency interpolator: the dimensions of the specified vectors "
-                    "are not consistent: start times (" +
-                    std::to_string( startTimes_.size( ) ) + "), end times (" + std::to_string( endTimes_.size( ) ) + "), ramp rates (" +
-                    std::to_string( rampRates_.size( ) ) + "), start frequencies (" + std::to_string( startFrequencies_.size( ) ) + ")." );
-        }
-
-        // Check if there are no discontinuities between end times and subsequent start times
-        for( unsigned int i = 1; i < startTimes_.size( ); ++i )
-        {
-            // If there are discontinuities
-            if( startTimes_.at( i ) != endTimes_.at( i - 1 ) )
-            {
-                //                // If the start and end times are inconsistent throw error
-                //                if ( endTimes_.at( i - 1 ) > startTimes_.at( i ) )
-                //                {
-                //                    throw std::runtime_error(
-                //                            "Error when creating piecewise linear frequency interpolator: inconsistency between ramp end "
-                //                            "time (" + std::to_string( double( endTimes_.at( i - 1 ) ) ) + ";" +
-                //                            basic_astrodynamics::getCalendarDateFromTime( endTimes_.at( i - 1 ) ).isoString( ) + ") and
-                //                            start time of the following ramp (" + std::to_string( double( startTimes_.at( i ) ) ) + ";" +
-                //                            basic_astrodynamics::getCalendarDateFromTime( startTimes_.at( i ) ).isoString( ) +
-                //                            "); the end is smaller than the start time." );
-                //                }
-                //                // If there are gaps in the data save that information
-                //                else
-                //                {
-                //                    invalidTimeBlocksStartTimes_.push_back( endTimes_.at( i - 1 ) );
-                //                    invalidTimeBlocksEndTimes_.push_back( startTimes_.at( i ) );
-                //                }
-            }
-        }
-
-        startTimeLookupScheme_ = std::make_shared< interpolators::HuntingAlgorithmLookupScheme< Time > >( startTimes_ );
-
-        if( !invalidTimeBlocksStartTimes_.empty( ) )
-        {
-            invalidStartTimeLookupScheme_ =
-                    std::make_shared< interpolators::BinarySearchLookupScheme< Time > >( invalidTimeBlocksStartTimes_ );
-        }
+        initialize( );
     }
 
+    void initialize( );
+
     /*! Templated function to compute the transmitted frequency at the specified time.
      *
      * Templated function to compute the transmitted frequency at the specified time. Frequency is computed according to
@@ -377,6 +336,8 @@ public:
         return startFrequencies_;
     }
 
+    void addFrequencyInterpolator( const std::shared_ptr< PiecewiseLinearFrequencyInterpolator > rampsToAdd );
+
 private:
     //! Get frequency (with long double as observation scalar type and double as time type).
     virtual double getCurrentFrequency( const double lookupTime )
diff --git a/include/tudat/io/readTrackingTxtFile.h b/include/tudat/io/readTrackingTxtFile.h
index cb4f59576..a348cea79 100644
--- a/include/tudat/io/readTrackingTxtFile.h
+++ b/include/tudat/io/readTrackingTxtFile.h
@@ -296,13 +296,13 @@ private:
      * Read out the raw data map from a filestream
      * @param dataFile filestream
      */
-    void readRawDataMap( std::ifstream& dataFile );
+    void readRawDataMap( std::ifstream& dataFile, const TrackingTxtFileReadFilterType dataFilterMethod = no_tracking_txt_file_filter );
 
     /*!
      * Process a single raw line from the file and add it to the data maps
      * @param rawLine single line from the file as a string
      */
-    void addLineToRawDataMap( std::string& rawLine );
+    void addLineToRawDataMap( std::string& rawLine, const TrackingTxtFileReadFilterType dataFilterMethod = no_tracking_txt_file_filter );
 
     bool validateCurrentLineProcessing( const TrackingTxtFileReadFilterType dataFilterMethod, const std::vector< std::string >& rawVector );
 
@@ -447,9 +447,10 @@ static inline std::shared_ptr< TrackingTxtFileContents > createTrackingTxtFileCo
                                                                                         const std::vector< std::string >& columnTypes,
                                                                                         char commentSymbol = '#',
                                                                                         const std::string& valueSeparators = ",: \t",
-                                                                                        const bool ignoreOmittedColumns = false )
+                                                                                        const bool ignoreOmittedColumns = false,
+                                                                                        const TrackingTxtFileReadFilterType dataFilterMethod = no_tracking_txt_file_filter )
 {
-    return std::make_shared< TrackingTxtFileContents >( fileName, columnTypes, commentSymbol, valueSeparators, ignoreOmittedColumns );
+    return std::make_shared< TrackingTxtFileContents >( fileName, columnTypes, commentSymbol, valueSeparators, ignoreOmittedColumns, dataFilterMethod );
 }
 
 inline std::shared_ptr< TrackingTxtFileContents > readIfmsFile( const std::string& fileName, const bool applyTroposphereCorrection = true )
@@ -467,7 +468,7 @@ inline std::shared_ptr< TrackingTxtFileContents > readIfmsFile( const std::strin
                                               "doppler_troposphere_correction",
                                               "doppler_noise_hz" } );
 
-    auto rawFileContents = createTrackingTxtFileContents( fileName, columnTypes, '#', ", \t", true );
+    auto rawFileContents = createTrackingTxtFileContents( fileName, columnTypes, '#', ", \t", true, ifms_tracking_txt_file_filter );
     rawFileContents->addMetaData( TrackingDataType::file_name, fileName );
     if( applyTroposphereCorrection )
     {
diff --git a/src/tudat/astro/ground_stations/transmittingFrequencies.cpp b/src/tudat/astro/ground_stations/transmittingFrequencies.cpp
index 78e926f6c..e04274b77 100644
--- a/src/tudat/astro/ground_stations/transmittingFrequencies.cpp
+++ b/src/tudat/astro/ground_stations/transmittingFrequencies.cpp
@@ -65,6 +65,80 @@ long double StationFrequencyInterpolator::getTemplatedFrequencyIntegral( const T
     return getLongFrequencyIntegral( quadratureStartTime, quadratureEndTime );
 }
 
+void PiecewiseLinearFrequencyInterpolator::initialize( )
+{
+    // Check if dimensions of all vectors are consistent
+    if( startTimes_.size( ) != endTimes_.size( ) || startTimes_.size( ) != rampRates_.size( ) ||
+        startTimes_.size( ) != startFrequencies_.size( ) )
+    {
+        throw std::runtime_error(
+                "Error when creating piecewise linear frequency interpolator: the dimensions of the specified vectors "
+                "are not consistent: start times (" +
+                std::to_string( startTimes_.size( ) ) + "), end times (" + std::to_string( endTimes_.size( ) ) + "), ramp rates (" +
+                std::to_string( rampRates_.size( ) ) + "), start frequencies (" + std::to_string( startFrequencies_.size( ) ) + ")." );
+    }
+
+    std::map< Time, Time > endTimesMap;
+    std::map< Time, double > rampRatesMap;
+    std::map< Time, double > startFrequenciesMap;
+
+    for( unsigned int i = 0; i < startTimes_.size( ); i++ )
+    {
+        endTimesMap[ startTimes_.at( i ) ] = endTimes_.at( i );
+        rampRatesMap[ startTimes_.at( i ) ] = rampRates_.at( i );
+        startFrequenciesMap[ startTimes_.at( i ) ] = startFrequencies_.at( i );
+    }
+
+    startTimes_ = utilities::createVectorFromMapKeys( endTimesMap );
+    endTimes_ = utilities::createVectorFromMapValues( endTimesMap );
+    rampRates_ = utilities::createVectorFromMapValues( rampRatesMap );
+    startFrequencies_ = utilities::createVectorFromMapValues( startFrequenciesMap );
+
+
+    startTimeLookupScheme_ = std::make_shared< interpolators::HuntingAlgorithmLookupScheme< Time > >( startTimes_ );
+
+    if( !invalidTimeBlocksStartTimes_.empty( ) )
+    {
+        invalidStartTimeLookupScheme_ =
+                std::make_shared< interpolators::BinarySearchLookupScheme< Time > >( invalidTimeBlocksStartTimes_ );
+    }
+}
+
+
+void PiecewiseLinearFrequencyInterpolator::addFrequencyInterpolator( const std::shared_ptr< PiecewiseLinearFrequencyInterpolator > rampsToAdd )
+{
+    std::vector< Time > startTimesToAdd = rampsToAdd->getStartTimes( );
+    std::vector< Time > endTimesToAdd = rampsToAdd->getEndTimes( );
+    std::vector< double > rampRatesToAdd = rampsToAdd->getRampRates( );
+    std::vector< double > startFrequenciesToAdd = rampsToAdd->getStartFrequencies( );
+
+    std::map< Time, Time > endTimesMap;
+    std::map< Time, double > rampRatesMap;
+    std::map< Time, double > startFrequenciesMap;
+
+    for( unsigned int i = 0; i < startTimes_.size( ); i++ )
+    {
+        endTimesMap[ startTimes_.at( i ) ] = endTimes_.at( i );
+        rampRatesMap[ startTimes_.at( i ) ] = rampRates_.at( i );
+        startFrequenciesMap[ startTimes_.at( i ) ] = startFrequencies_.at( i );
+    }
+
+    for( unsigned int i = 0; i < startTimesToAdd.size( ); i++ )
+    {
+        endTimesMap[ startTimesToAdd.at( i ) ] = endTimesToAdd.at( i );
+        rampRatesMap[ startTimesToAdd.at( i ) ] = rampRatesToAdd.at( i );
+        startFrequenciesMap[ startTimesToAdd.at( i ) ] = startFrequenciesToAdd.at( i );
+    }
+
+    startTimes_ = utilities::createVectorFromMapKeys( endTimesMap );
+    endTimes_ = utilities::createVectorFromMapValues( endTimesMap );
+    rampRates_ = utilities::createVectorFromMapValues( rampRatesMap );
+    startFrequencies_ = utilities::createVectorFromMapValues( startFrequenciesMap );
+
+    initialize( );
+
+}
+
 }  // namespace ground_stations
 
 }  // namespace tudat
diff --git a/src/tudat/io/readTrackingTxtFile.cpp b/src/tudat/io/readTrackingTxtFile.cpp
index 5d97cf93f..b858a995a 100644
--- a/src/tudat/io/readTrackingTxtFile.cpp
+++ b/src/tudat/io/readTrackingTxtFile.cpp
@@ -22,11 +22,11 @@ void TrackingTxtFileContents::parseData( const TrackingTxtFileReadFilterType dat
     {
         throw std::runtime_error( "Error when opening Tracking txt file: file " + fileName_ + " could not be opened." );
     }
-    readRawDataMap( dataFile );
+    readRawDataMap( dataFile, dataFilterMethod );
     convertDataMap( dataFilterMethod );
 }
 
-void TrackingTxtFileContents::readRawDataMap( std::ifstream& dataFile )
+void TrackingTxtFileContents::readRawDataMap( std::ifstream& dataFile, const TrackingTxtFileReadFilterType dataFilterMethod  )
 {
     std::string currentLine;
 
@@ -34,12 +34,67 @@ void TrackingTxtFileContents::readRawDataMap( std::ifstream& dataFile )
     {
         if( !currentLine.empty( ) && currentLine.at( 0 ) != commentSymbol_ )
         {
-            addLineToRawDataMap( currentLine );
+            addLineToRawDataMap( currentLine, dataFilterMethod );
         }
     }
 }
 
-void TrackingTxtFileContents::addLineToRawDataMap( std::string& rawLine )
+
+bool isIfmsEntryInvalid( const std::string ifmsFileEntry )
+{
+    if( ifmsFileEntry.size( ) == 0 )
+    {
+        return false;
+    }
+    else
+    {
+        if( ifmsFileEntry.at( 0 ) != '-' )
+        {
+            return true;
+        }
+        else
+        {
+            bool seenNonNine = false; bool seenDot = false;
+            for( unsigned int i = 1; i < ifmsFileEntry.size( ); i++ )
+            {
+                if( ifmsFileEntry.at( i ) == '.' )
+                {
+                    seenDot = true;
+                }
+                else if( ifmsFileEntry.at( i ) != '9' )
+                {
+                    seenNonNine = true;
+                }
+            }
+            return !( seenDot == true && seenNonNine == false );
+        }
+    }
+}
+
+bool TrackingTxtFileContents::validateCurrentLineProcessing( const TrackingTxtFileReadFilterType dataFilterMethod,
+                                                             const std::vector< std::string >& rawVector )
+{
+    bool addLine = true;
+
+    if( dataFilterMethod == ifms_tracking_txt_file_filter )
+    {
+        for( unsigned int i = 0; i < rawVector.size( ); i++ )
+        {
+            if( ( i == 6 ) || ( i == 8 ) || ( i == 10 ) )
+            {
+                addLine = isIfmsEntryInvalid( rawVector.at( i ) );
+                if( !addLine )
+                {
+                    return addLine;
+                }
+            }
+        }
+    }
+
+    return addLine;
+}
+
+void TrackingTxtFileContents::addLineToRawDataMap( std::string& rawLine, const TrackingTxtFileReadFilterType dataFilterMethod )
 {
     size_t numberOfColumns = getNumColumns( );
     std::vector< std::string > currentSplitRawLine_;
@@ -63,49 +118,18 @@ void TrackingTxtFileContents::addLineToRawDataMap( std::string& rawLine )
         }
     }
 
-    // Populate the dataMap_ with a new row on each of the vectors
-    for( std::size_t i = 0; i < numberOfColumns; ++i )
-    {
-        std::string currentFieldType = columnFieldTypes_.at( i );
-        std::string currentValue = currentSplitRawLine_.at( i );
-        rawDataMap_[ currentFieldType ].push_back( currentValue );
-    }
-}
 
-bool TrackingTxtFileContents::validateCurrentLineProcessing( const TrackingTxtFileReadFilterType dataFilterMethod,
-                                                             const std::vector< std::string >& rawVector )
-{
-    bool addLine = true;
 
-    if( dataFilterMethod == ifms_tracking_txt_file_filter )
+    // Populate the dataMap_ with a new row on each of the vectors
+    if( validateCurrentLineProcessing( dataFilterMethod, currentSplitRawLine_ ) )
     {
-        for( unsigned int i = 0; i < rawVector.size( ); i++ )
+        for( std::size_t i = 0; i < numberOfColumns; ++i )
         {
-            bool currentEntryIsValid = false;
-            if( rawVector.at( 0 ) != "-" )
-            {
-                currentEntryIsValid = true;
-            }
-            else
-            {
-                for( unsigned int j = 1; j < rawVector.at( j ).size( ); j++ )
-                {
-                    if( ( rawVector.at( j ) != "9" ) && ( rawVector.at( j ) != "." ) )
-                    {
-                        currentEntryIsValid = true;
-                        break;
-                    }
-                }
-            }
-            if( currentEntryIsValid == false )
-            {
-                addLine = false;
-                break;
-            }
+            std::string currentFieldType = columnFieldTypes_.at( i );
+            std::string currentValue = currentSplitRawLine_.at( i );
+            rawDataMap_[ currentFieldType ].push_back( currentValue );
         }
     }
-
-    return addLine;
 }
 
 void TrackingTxtFileContents::convertDataMap( const TrackingTxtFileReadFilterType dataFilterMethod )
@@ -128,15 +152,12 @@ void TrackingTxtFileContents::convertDataMap( const TrackingTxtFileReadFilterTyp
         const TrackingDataType& dataType = converter->getTrackingDataType( );
 
         // Convert and store double vector
-        if( validateCurrentLineProcessing( dataFilterMethod, rawVector ) )
+        std::vector< double > dataVector;
+        for( std::string rawValue: rawVector )
         {
-            std::vector< double > dataVector;
-            for( std::string rawValue: rawVector )
-            {
-                dataVector.push_back( converter->toDouble( rawValue ) );
-            }
-            doubleDataMap_[ dataType ] = dataVector;
+            dataVector.push_back( converter->toDouble( rawValue ) );
         }
+        doubleDataMap_[ dataType ] = dataVector;
     }
 }
 
diff --git a/src/tudat/simulation/estimation_setup/processTrackingTxtFile.cpp b/src/tudat/simulation/estimation_setup/processTrackingTxtFile.cpp
index 7ef62b48e..5ab525530 100644
--- a/src/tudat/simulation/estimation_setup/processTrackingTxtFile.cpp
+++ b/src/tudat/simulation/estimation_setup/processTrackingTxtFile.cpp
@@ -82,14 +82,7 @@ void setStationFrequenciesFromTrackingData(
                 throw std::runtime_error( "Error when reading IFMS transmitter frequencies, only unramped data currently supported." );
             }
         }
-        rampStartTimes[ 0 ] -= 1.0;
-        for( unsigned int i = 0; i < rampStartTimes.size( ) - 1; i++ )
-        {
-            double timeDifference = rampStartTimes.at( i + 1 ) - rampEndTimes.at( i );
-            rampStartTimes[ i + 1 ] -= timeDifference / 2.0;
-            rampEndTimes[ i ] += timeDifference / 2.0;
-        }
-        rampEndTimes[ rampEndTimes.size( ) - 1 ] += 1.0;
+
         rampInterpolators[ it.first ] = std::make_shared< ground_stations::PiecewiseLinearFrequencyInterpolator >(
                 rampStartTimes, rampEndTimes, rampRates, rampStartFrequencies );
     }
@@ -100,7 +93,22 @@ void setStationFrequenciesFromTrackingData(
         {
             throw std::runtime_error( "Error when setting frequencies for station " + it->first + ", station not found." );
         }
-        bodies.at( "Earth" )->getGroundStation( it->first )->setTransmittingFrequencyCalculator( it->second );
+        if( !bodies.at( "Earth" )->getGroundStation( it->first )->hasFrequencyCalculator( ) )
+        {
+            bodies.at( "Earth" )->getGroundStation( it->first )->setTransmittingFrequencyCalculator( it->second );
+        }
+        else if( std::dynamic_pointer_cast< ground_stations::PiecewiseLinearFrequencyInterpolator >(
+                         bodies.at( "Earth" )->getGroundStation( it->first )->getTransmittingFrequencyCalculator( ) ) != nullptr )
+        {
+            std::shared_ptr< ground_stations::PiecewiseLinearFrequencyInterpolator > existingFrequencyInterpolator =
+                std::dynamic_pointer_cast< ground_stations::PiecewiseLinearFrequencyInterpolator >(
+                    bodies.at( "Earth" )->getGroundStation( it->first )->getTransmittingFrequencyCalculator( ) );
+            existingFrequencyInterpolator->addFrequencyInterpolator( it->second );
+        }
+        else
+        {
+            throw std::runtime_error( "Error when adding ramp tables for station " + it->first + ", existing frequency calculator implemented, but not of correct type" );
+        }
     }
 }
 
diff --git a/src/tudatpy/data/expose_data.cpp b/src/tudatpy/data/expose_data.cpp
index c4f2a1baa..040ae48e4 100644
--- a/src/tudatpy/data/expose_data.cpp
+++ b/src/tudatpy/data/expose_data.cpp
@@ -321,6 +321,12 @@ void expose_data( py::module& m )
             .value( "scan_nr", tudat::input_output::TrackingDataType::scan_nr )
             .export_values( );
 
+    py::enum_< tudat::input_output::TrackingTxtFileReadFilterType >( m, "TrackingTxtFileReadFilterType", R"doc(No documentation available.)doc" )
+            .value( "no_tracking_txt_file_filter", tudat::input_output::TrackingTxtFileReadFilterType::no_tracking_txt_file_filter, R"doc(No documentation available.)doc" )
+            .value( "ifms_tracking_txt_file_filter", tudat::input_output::TrackingTxtFileReadFilterType::ifms_tracking_txt_file_filter, R"doc(No documentation available.)doc" )
+            .export_values( );
+
+
     py::class_< tio::solar_activity::SolarActivityData, std::shared_ptr< tio::solar_activity::SolarActivityData > >(
             m, "SolarActivityData", R"doc(No documentation available.)doc" )
             .def_readonly( "solar_radio_flux_107_observed", &tio::solar_activity::SolarActivityData::solarRadioFlux107Observed );
@@ -520,7 +526,8 @@ void expose_data( py::module& m )
            py::arg( "column_types" ),
            py::arg( "comment_symbol" ) = '#',
            py::arg( "value_separators" ) = ",:\t ",
-           py::arg( "ignore_omitted_columns" ) = false );
+           py::arg( "ignore_omitted_columns" ) = false,
+           py::arg( "data_filter_method" ) = tio::no_tracking_txt_file_filter);
 
     m.def( "grail_antenna_file_reader", &tio::grailAntennaFileReader, py::arg( "file_name" ), R"doc(No documentation available.)doc" );
     m.def( "grail_mass_level_0_file_reader",
diff --git a/tests/test_tudat/data.zip b/tests/test_tudat/data.zip
index 0cb934d7b..6dd089e7d 100644
Binary files a/tests/test_tudat/data.zip and b/tests/test_tudat/data.zip differ
diff --git a/tests/test_tudat/src/io/unitTestIfmsFileReader.cpp b/tests/test_tudat/src/io/unitTestIfmsFileReader.cpp
index afc49fa73..f8f2d1201 100644
--- a/tests/test_tudat/src/io/unitTestIfmsFileReader.cpp
+++ b/tests/test_tudat/src/io/unitTestIfmsFileReader.cpp
@@ -30,6 +30,8 @@ using namespace tudat::input_output;
 using namespace tudat::basic_astrodynamics;
 using namespace tudat::observation_models;
 using namespace tudat::simulation_setup;
+using namespace tudat::spice_interface;
+using namespace tudat::ground_stations;
 using namespace tudat;
 
 namespace tudat
@@ -72,6 +74,8 @@ BOOST_AUTO_TEST_CASE( testIfmsFileReader )
     Eigen::Vector3d stationPosition = getCombinedApproximateGroundStationPositions( ).at( "NWNORCIA" );
     auto timeConverter = earth_orientation::createDefaultTimeConverter( );
 
+    std::vector< int > linesToBeSkipped = { 1, 1, 0 };
+
     for( unsigned int i = 0; i < rawIfmsFiles.size( ) + 1; i++ )
     {
         std::vector< std::shared_ptr< ProcessedTrackingTxtFileContents< double, Time > > > processedIfmsFiles;
@@ -101,40 +105,178 @@ BOOST_AUTO_TEST_CASE( testIfmsFileReader )
 
         if( i < rawIfmsFiles.size( ) )
         {
-            BOOST_CHECK_EQUAL( observationValues.rows( ), rawIfmsFileSizes.at( i ) );
-            BOOST_CHECK_CLOSE_FRACTION( observationValues( 9 ), tenthObservation.at( i ), std::numeric_limits< double >::epsilon( ) );
-            Time utcTime = timeConverter->getCurrentTime< Time >( tdb_scale, utc_scale, observationTimes.at( 9 ), stationPosition );
+            BOOST_CHECK_EQUAL( observationValues.rows( ), rawIfmsFileSizes.at( i ) - linesToBeSkipped.at( i ) );
+            BOOST_CHECK_CLOSE_FRACTION( observationValues( 9 - linesToBeSkipped.at( i ) ), tenthObservation.at( i ), std::numeric_limits< double >::epsilon( ) );
+            Time utcTime = timeConverter->getCurrentTime< Time >( tdb_scale, utc_scale, observationTimes.at( 9  - linesToBeSkipped.at( i ) ), stationPosition );
             Time utcTimeTest = timeFromIsoString< Time >( tenthObservationTimeUtc.at( i ) );
             BOOST_CHECK_SMALL( static_cast< double >( utcTime - utcTimeTest ), 1.0E-12 );
         }
         else
         {
-            BOOST_CHECK_EQUAL( observationValues.rows( ), rawIfmsFileSizes.at( 0 ) + rawIfmsFileSizes.at( 1 ) + rawIfmsFileSizes.at( 2 ) );
-            BOOST_CHECK_CLOSE_FRACTION( observationValues( 9 ), tenthObservation.at( 0 ), std::numeric_limits< double >::epsilon( ) );
-            BOOST_CHECK_CLOSE_FRACTION( observationValues( 9 + rawIfmsFileSizes.at( 0 ) ),
+            BOOST_CHECK_EQUAL( observationValues.rows( ), rawIfmsFileSizes.at( 0 ) + rawIfmsFileSizes.at( 1 ) + rawIfmsFileSizes.at( 2 ) -
+                                       linesToBeSkipped.at( 0 ) - linesToBeSkipped.at( 1 ) - linesToBeSkipped.at( 2 ) );
+            BOOST_CHECK_CLOSE_FRACTION( observationValues( 9 - linesToBeSkipped.at( 0 ) ), tenthObservation.at( 0 ), std::numeric_limits< double >::epsilon( ) );
+            BOOST_CHECK_CLOSE_FRACTION( observationValues( 9 + rawIfmsFileSizes.at( 0 ) - linesToBeSkipped.at( 0 ) - linesToBeSkipped.at( 1 ) ),
                                         tenthObservation.at( 1 ),
                                         std::numeric_limits< double >::epsilon( ) );
-            BOOST_CHECK_CLOSE_FRACTION( observationValues( 9 + rawIfmsFileSizes.at( 0 ) + rawIfmsFileSizes.at( 1 ) ),
+            BOOST_CHECK_CLOSE_FRACTION( observationValues( 9 + rawIfmsFileSizes.at( 0 ) + rawIfmsFileSizes.at( 1 )
+                                                           - linesToBeSkipped.at( 0 ) - linesToBeSkipped.at( 1 ) - linesToBeSkipped.at( 2 ) ),
                                         tenthObservation.at( 2 ),
                                         std::numeric_limits< double >::epsilon( ) );
 
-            Time utcTime1 = timeConverter->getCurrentTime< Time >( tdb_scale, utc_scale, observationTimes.at( 9 ), stationPosition );
+            Time utcTime1 = timeConverter->getCurrentTime< Time >( tdb_scale, utc_scale, observationTimes.at( 9 - linesToBeSkipped.at( 0 ) ), stationPosition );
             Time utcTimeTest1 = timeFromIsoString< Time >( tenthObservationTimeUtc.at( 0 ) );
             BOOST_CHECK_SMALL( static_cast< double >( utcTime1 - utcTimeTest1 ), 1.0E-12 );
 
             Time utcTime2 = timeConverter->getCurrentTime< Time >(
-                    tdb_scale, utc_scale, observationTimes.at( 9 + rawIfmsFileSizes.at( 0 ) ), stationPosition );
+                    tdb_scale, utc_scale, observationTimes.at( 9 + rawIfmsFileSizes.at( 0 ) - linesToBeSkipped.at( 0 ) - linesToBeSkipped.at( 1 )  ), stationPosition );
             Time utcTimeTest2 = timeFromIsoString< Time >( tenthObservationTimeUtc.at( 1 ) );
             BOOST_CHECK_SMALL( static_cast< double >( utcTime2 - utcTimeTest2 ), 1.0E-12 );
 
             Time utcTime3 = timeConverter->getCurrentTime< Time >(
-                    tdb_scale, utc_scale, observationTimes.at( 9 + rawIfmsFileSizes.at( 0 ) + rawIfmsFileSizes.at( 1 ) ), stationPosition );
+                    tdb_scale, utc_scale, observationTimes.at( 9 + rawIfmsFileSizes.at( 0 ) + rawIfmsFileSizes.at( 1 ) - linesToBeSkipped.at( 0 ) - linesToBeSkipped.at( 1 ) - linesToBeSkipped.at( 2 ) ), stationPosition );
             Time utcTimeTest3 = timeFromIsoString< Time >( tenthObservationTimeUtc.at( 2 ) );
             BOOST_CHECK_SMALL( static_cast< double >( utcTime3 - utcTimeTest3 ), 1.0E-12 );
         }
     }
 }
 
+std::map< Time, double > loadIfmsFilesCombined(const std::vector< std::string >& ifmsFileNames,
+                            SystemOfBodies& bodies)
+{
+    std::map< Time, double > uplinkFrequencies;
+
+    // Build a single observation collection from all IFMS files
+    std::shared_ptr< ObservationCollection< double, Time > > observationCollectionCombined =
+            createIfmsObservedObservationCollectionFromFiles< double, Time >(
+                    ifmsFileNames,
+                    bodies,
+                    "spacecraft",
+                    "NWNORCIA", FrequencyBands::x_band, FrequencyBands::x_band );
+
+    // Retrieve frequency calculator
+    std::shared_ptr< ground_stations::GroundStation > nwnorcia =
+            bodies.at( "Earth" )->getGroundStation( "NWNORCIA" );
+    std::shared_ptr< StationFrequencyInterpolator > freqCalc =
+            nwnorcia->getTransmittingFrequencyCalculator( );
+
+    // Extract concatenated observation times and compute frequencies
+    std::vector< Time > epochs =
+            observationCollectionCombined->getConcatenatedObservationTimes();
+    for (double epoch : epochs)
+    {
+        uplinkFrequencies[ epoch ] = freqCalc->getTemplatedCurrentFrequency( epoch ) ;
+    }
+    return uplinkFrequencies;
+}
+
+std::map< Time, double > loadIfmsFilesSeparate(const std::vector< std::string >& ifmsFileNames,
+                          SystemOfBodies& bodies)
+{
+    std::map< Time, double > uplinkFrequencies;
+
+    // Build a separate observation collection from each IFMS file
+    std::vector< std::shared_ptr< ObservationCollection< double, Time > > > observationCollections;
+    for (const std::string& fileName : ifmsFileNames)
+    {
+        observationCollections.push_back( createIfmsObservedObservationCollectionFromFiles< double, Time >(
+                        std::vector< std::string >({fileName}),
+                        bodies,
+                        "spacecraft",
+                        "NWNORCIA", FrequencyBands::x_band, FrequencyBands::x_band ) );
+    }
+
+    // Retrieve frequency calculator
+    std::shared_ptr< ground_stations::GroundStation > nwnorcia =
+            bodies.at( "Earth" )->getGroundStation( "NWNORCIA" );
+    std::shared_ptr< StationFrequencyInterpolator > freqCalc =
+            nwnorcia->getTransmittingFrequencyCalculator();
+
+    for( unsigned int i = 0; i < observationCollections.size( ); i++ )
+    {
+        // Extract concatenated observation times and compute frequencies
+        std::vector< Time > epochs =
+                observationCollections.at( i )->getConcatenatedObservationTimes();
+        for (double epoch : epochs)
+        {
+            uplinkFrequencies[ epoch ] = freqCalc->getTemplatedCurrentFrequency( epoch ) ;
+        }
+    }
+    return uplinkFrequencies;
+}
+
+BOOST_AUTO_TEST_CASE( testMultipleIfmsFileReader )
+{
+    // Adjust this to point to the directory containing your IFMS files
+    const std::string SOURCE_DIR = paths::getTudatTestDataPath( ) + "/estrack_n_way_doppler_observation_model/";
+
+    // Define IFMS file paths
+    std::vector< std::string > ifmsFilesOrdered;
+    ifmsFilesOrdered.push_back( SOURCE_DIR + "M32ICL1L02_D2X_133612305_00.TAB" );
+    ifmsFilesOrdered.push_back( SOURCE_DIR + "M32ICL1L02_D2X_133621819_00.TAB" );
+    ifmsFilesOrdered.push_back( SOURCE_DIR + "M32ICL1L02_D2X_133630120_00.TAB" );
+
+    std::vector< std::string > ifmsFilesUnordered;
+    ifmsFilesUnordered.push_back( ifmsFilesOrdered.at( 2 ) );
+    ifmsFilesUnordered.push_back( ifmsFilesOrdered.at( 1 ) );
+    ifmsFilesUnordered.push_back( ifmsFilesOrdered.at( 0 ) );
+
+
+    std::vector< std::map< Time, double > > transmittedFrequencies;
+    for( int test = 0; test < 4; test++ )
+    {
+        spice_interface::loadStandardSpiceKernels( );
+
+        // Create bodies
+        std::vector< std::string > bodiesToCreate;
+        bodiesToCreate.push_back( "Earth" );
+        BodyListSettings bodySettings =
+                getDefaultBodySettings(
+                        bodiesToCreate,
+                        "SSB",
+                        "J2000" );
+        bodySettings.at( "Earth" )->groundStationSettings = getRadioTelescopeStationSettings( );
+        bodySettings.addSettings( "spacecraft" );
+        SystemOfBodies bodies = createSystemOfBodies( bodySettings );
+
+        // Load files separately or one by one
+        if( test == 0 )
+        {
+            transmittedFrequencies.push_back( loadIfmsFilesCombined( ifmsFilesOrdered, bodies ) );
+        }
+        else if( test == 1 )
+        {
+            transmittedFrequencies.push_back( loadIfmsFilesSeparate( ifmsFilesOrdered, bodies ) );
+        }
+        else if( test == 2 )
+        {
+            transmittedFrequencies.push_back( loadIfmsFilesCombined( ifmsFilesUnordered, bodies ) );
+        }
+        else if( test == 3 )
+        {
+            transmittedFrequencies.push_back( loadIfmsFilesSeparate( ifmsFilesUnordered, bodies ) );
+        }
+    }
+
+    BOOST_CHECK_EQUAL( transmittedFrequencies.at( 0 ).size( ), transmittedFrequencies.at( 1 ).size( ) );
+    BOOST_CHECK_EQUAL( transmittedFrequencies.at( 0 ).size( ), transmittedFrequencies.at( 2 ).size( ) );
+    BOOST_CHECK_EQUAL( transmittedFrequencies.at( 0 ).size( ), transmittedFrequencies.at( 3 ).size( ) );
+
+    for( auto it : transmittedFrequencies.at( 0 ) )
+    {
+        BOOST_CHECK_CLOSE_FRACTION( transmittedFrequencies.at( 0 ).at( it.first ),
+                                    transmittedFrequencies.at( 1 ).at( it.first ),
+                                    10.0 * std::numeric_limits< double >::epsilon( ) );
+        BOOST_CHECK_CLOSE_FRACTION( transmittedFrequencies.at( 0 ).at( it.first ),
+                                    transmittedFrequencies.at( 2 ).at( it.first ),
+                                    10.0 * std::numeric_limits< double >::epsilon( ) );
+        BOOST_CHECK_CLOSE_FRACTION( transmittedFrequencies.at( 0 ).at( it.first ),
+                                    transmittedFrequencies.at( 3 ).at( it.first ),
+                                    10.0 * std::numeric_limits< double >::epsilon( ) );
+    }
+}
+
+
 // End test suite
 BOOST_AUTO_TEST_SUITE_END( );
 
